core/state-machine/src/changes_trie/build_cache.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Changes tries build cache.
+
+use std::collections::{HashMap, HashSet};
+
+/// Changes trie build cache.
+///
+/// Helps to avoid read of changes tries from the database when digest trie
+/// is built. It holds changed keys for every block (indexed by changes trie
+/// root) that could be referenced by future digest items. For digest entries
+/// it also holds keys covered by this digest.
+///
+/// Entries are pruned from the cache once digest block that is using this entry
+/// is inserted (because digest block will includes all keys from this entry).
+/// When there's a fork, entries are pruned when first changes trie is inserted.
+pub struct BuildCache<H, N> {
+	/// Map of block (implies changes true) number => changes trie root.
+	roots_by_number: HashMap<N, H>,
```

Comment:
I wonder if it could be remove by returning the hash from the digest instead of their number in 'prepare_input'.

Reply:
It actually plays slightly different role - there could be only one entry per block **number**. So if we'll get rid of numbers at all, we need some other strategies to prune the cache. That's what I was trying to avoid - that is mentioned in description: "but it'll make it complex + memory demanding". IMO it is better to stick now with the strategy that will be super-fast when there are no competing forks, than to strategy that will be simply faster when there are competing forks. If we'll see situations when there are many competing forks, we could change the impl, or introduce second mode.

Reply:
Ok, I start to understand how it works with fork :), the pruning logic works as long as the same block indexes (number) are used for CT digest building in every fork.

Is there a possible issue with ConfigurationRange (different range for building blocks in two forks then some blocks are not freed, but should be fairly marginal).

Reply:
When there are two competing forks:
1) the last fork that inserts data into cache, will win;
2) if there are configuration changes in different forks - that would lead to `CacheAction::Clear`. It isn't super-effective, but as I said, the cache is optimized to work in non-competing environment.

Reply:
Hmmm...OK, I see, there's a possible issue:
1) initially there are no forks && config is 4^2;
2) then the fork appears that changes configuration to 1^1 (no digests) - the cache is cleared;
3) some blocks for original forks are created => we're adding corresponding entries;
4) then the fork1 wins (i.e. the block from fork1 is finalized) => we'll never prune entries from (3) until configuration will be changed again.

I believe the fix should be: when top-level digest is created, clear the cache. This isn't, again, super-effective for competing forks, but should be fine for single-fork environments.

Reply:
I've fixed that by clearing cache when: digests are disabled, top-level digest block is built, or skewed digest is built.